// mojio-js/lib/nodejs/SignalRNodeWrapper.js
// Generated by CoffeeScript 1.6.3
(function() {
  var SignalR, SignalRNodeWrapper,
    __bind = function(fn, me){ return function(){ return fn.apply(me, arguments); }; };

  SignalR = require('signalr-client');

  module.exports = SignalRNodeWrapper = (function() {
    SignalRNodeWrapper.prototype.observer_callbacks = {};

    SignalRNodeWrapper.prototype.observer_registry = function(entity) {
      var callback, _i, _len, _ref, _results;
      if (this.observer_callbacks[entity._id]) {
        _ref = this.observer_callbacks[entity._id];
        _results = [];
        for (_i = 0, _len = _ref.length; _i < _len; _i++) {
          callback = _ref[_i];
          _results.push(callback(entity));
        }
        return _results;
      }
    };

    function SignalRNodeWrapper(url, hubNames, jquery) {
      this.observer_registry = __bind(this.observer_registry, this);
      this.url = url;
      this.hubs = {};
      this.signalr = new SignalR.client(url, hubNames, null);
    }

    SignalRNodeWrapper.prototype.getHub = function(which, callback) {
      if (this.hubs[which] != null) {
        return callback(null, this.hubs[which]);
      } else {
        this.hubs[which] = this.signalr.hub(which);
        this.hubs[which].on("Error", function(data) {
          return log(data);
        });
        this.hubs[which].on("UpdateEntity", this.observer_registry);
        return callback(null, this.hubs[which]);
      }
    };

    SignalRNodeWrapper.prototype.setCallback = function(id, futureCallback) {
      if (this.observer_callbacks[id] == null) {
        this.observer_callbacks[id] = [];
      }
      this.observer_callbacks[id].push(futureCallback);
    };

    SignalRNodeWrapper.prototype.removeCallback = function(id, pastCallback) {
      var callback, temp, _i, _len, _ref;
      if (pastCallback === null) {
        this.observer_callbacks[id] = [];
      } else {
        temp = [];
        _ref = this.observer_callbacks[id];
        for (_i = 0, _len = _ref.length; _i < _len; _i++) {
          callback = _ref[_i];
          if (callback !== pastCallback) {
            temp.push(callback);
          }
        }
        this.observer_callbacks[id] = temp;
      }
    };

    SignalRNodeWrapper.prototype.subscribe = function(hubName, method, subject, object, futureCallback, callback) {
      this.setCallback(subject, futureCallback);
      return this.getHub(hubName, function(error, hub) {
        if (error != null) {
          callback(error, null);
        }
        if (hub != null) {
          hub.invoke(method, object);
        }
        return callback(null, hub);
      });
    };

    SignalRNodeWrapper.prototype.unsubscribe = function(hubName, method, subject, object, pastCallback, callback) {
      this.removeCallback(subject, pastCallback);
      if (this.observer_callbacks[subject].length === 0) {
        return this.getHub(hubName, function(error, hub) {
          if (error != null) {
            callback(error, null);
          }
          if (hub != null) {
            hub.invoke(method, object);
          }
          return callback(null, hub);
        });
      } else {
        return callback(null, true);
      }
    };

    return SignalRNodeWrapper;

  })();

}).call(this);
